### PR TITLE
feat: contract fixture

### PR DESCRIPTION
**Pull request by Ninjagod1251**
### What I did

Started the Contract fixture implementation

fixes: #1241 

### Checklist

- [ ] All changes are completed
- [ ] New test cases have been added
- [ ] Documentation has been updated

### CODE REVIEW

File: src/ape/pytest/fixtures.py [hunk 1]
```diff
@@ -76,6 +77,13 @@ def project(self) -> ProjectManager:
 
         return self.project_manager
 
+    @pytest.fixture(scope="session")
+    def on_chain_contract(ContractInstance):
+        """
+        Access contract types and dependencies.
+        """
+        return ContractInstance.address()
```
Comment (fubuloubu) [inline src/ape/pytest/fixtures.py:85]: Look at where `ape.Contract` comes from

File: src/ape/pytest/fixtures.py [hunk 2]
```diff
@@ -76,6 +77,13 @@ def project(self) -> ProjectManager:
 
         return self.project_manager
 
+    @pytest.fixture(scope="session")
+    def on_chain_contract(ContractInstance):
```
Comment (antazoey) [inline src/ape/pytest/fixtures.py:81]: arg name should be `self` not sure why you put `ContractInstance` here tbh

File: src/ape/pytest/fixtures.py (hunk 1 above)
Comment (antazoey) [inline src/ape/pytest/fixtures.py:85]: i dont think there is a class method `address` for ContractInstance, but if there was, that is not what we want here.

Return `self.chain_manager.instance_at` the method, the return type is a callable that takes an address and returns a `ContractInstance`

File: tests/conftest.py [hunk 3]
```diff
@@ -77,6 +77,11 @@ def project_folder():
     return PROJECT_FOLDER
 
 
+@pytest.fixture(scope="session")
```
Comment (antazoey) [inline tests/conftest.py:80]: ?

**Review by antazoey — COMMENTED** [review]
@NotPeopling2day is helping!!

File: src/ape/pytest/fixtures.py [hunk 4]
```diff
@@ -76,6 +76,13 @@ def project(self) -> ProjectManager:
 
         return self.project_manager
 
+    @pytest.fixture(scope="session")
+    def on_chain_contract(self):
```
Comment (antazoey) [inline src/ape/pytest/fixtures.py:80]: this need to be called `Contract` to align with `ape.Contract`

File: src/ape/pytest/fixtures.py [hunk 5]
```diff
@@ -76,6 +76,13 @@ def project(self) -> ProjectManager:
...
+        """
+        Access contract types and dependencies.
```
Comment (antazoey) [inline src/ape/pytest/fixtures.py:82]:
```suggestion
        Instantiate a reference to an on-chain contract using its address (same as ``ape.Contract``).
```

**Review by antazoey — COMMENTED** [review]
need test, see other fixture tests

File: tests/functional/test_contract_instance.py [hunk 6]
```diff
@@ -54,6 +54,11 @@ def test_contract_calls(owner, contract_instance):
     assert contract_instance.myNumber() == 2
 
 
+def test_contract_fixture(contract_instance):
```
Comment (antazoey) [inline tests/functional/test_contract_instance.py:57]: no this is not testing the fixture, you need to add an integration test.
add a test to this file instead: https://github.com/ApeWorX/ape/blob/main/tests/integration/cli/projects/test/tests/test_fixtures.py

File: src/ape/pytest/fixtures.py [hunk 7]
```diff
@@ -76,6 +76,13 @@ def project(self) -> ProjectManager:
 
         return self.project_manager
 
+    @pytest.fixture(scope="session")
+    def contract(self):
```
Comment (antazoey) [inline src/ape/pytest/fixtures.py:80]: do capital C

File: tests/functional/test_contract_instance.py [hunk 8]
```diff
@@ -26,6 +27,11 @@ class DataObject(BaseModel):
     return DataObject()
 
 
+def test_use_contract(Contract, address):
```
Comment (antazoey) [inline tests/functional/test_contract_instance.py:30]: ape fixtures are not available in core's tests like this, do remove this and only use the integration test

File: tests/integration/cli/projects/test/tests/test_fixtures.py [hunk 9]
```diff
@@ -12,6 +12,10 @@ def test_chain(chain):
     assert isinstance(chain, ChainManager)
 
 
+def test_contract(Contract):
+    assert isinstance(Contract, ContractInstance)
```
Comment (antazoey) [inline tests/integration/cli/projects/test/tests/test_fixtures.py:16]: no this is callable, this test will fail (if you run it)...
you need to get an address from somewhere else and then `Contract(address)` will be the instance

**Comment by antazoey** [comment]
probably should also mention this feature in the testing user guide

File: tests/integration/cli/projects/test/tests/test_fixtures.py [hunk 10]
```diff
@@ -12,8 +13,10 @@ def test_chain(chain):
     assert isinstance(chain, ChainManager)
 
 
-def test_contract(Contract):
-    assert isinstance(Contract, ContractInstance)
+def test_contract(accounts):
```
Comment (antazoey) [inline tests/integration/cli/projects/test/tests/test_fixtures.py]: I dont see the `Contract` being used?
That is the fixture you are adding, that is what you should be testing.

**Review by antazoey — CHANGES_REQUESTED** [review]
test does not test the newly added code.

File: tests/integration/cli/projects/test/tests/test_fixtures.py [hunk 11]
```diff
@@ -12,6 +12,10 @@ def test_chain(chain):
     assert isinstance(chain, ChainManager)
 
 
+def test_contract(chain, Contract):
+    assert chain.contracts.instance_at == Contract
```
Comment (antazoey) [inline tests/integration/cli/projects/test/tests/test_fixtures.py:16]: actual goes first

```suggestion
    assert Contract == chain.contracts.instance_at
```

**Review by antazoey — DISMISSED** [review]

**Comment by antazoey** [comment]
Thanks Ninja god, this is an excellent PR!

**Review by fubuloubu — APPROVED** [review]

File: tests/integration/cli/projects/test/tests/test_fixtures.py [hunk 12]
```diff
@@ -1,3 +1,4 @@
+from ape.contracts import ContractInstance
```
Comment (antazoey) [inline tests/integration/cli/projects/test/tests/test_fixtures.py:1]: my bad, was trying to fix the commit history but messed it up

```suggestion
```

**Pull request merged**